### PR TITLE
Deprecate all the fetch* for a more modern/perl6? approch with a row and allrows method

### PR DESCRIPTION
**Pull request by Skarsnik**
Ignore the title.
Added a row and that return typed value when possible: a int field in the DB will be an Int.
Tagged fetch_typedhash as deprecated

Close:
#31 #22 (maybe?)

TODO:
Probably rewrite Pg _row to use typed C call and not convert from Str.

### CODE REVIEW

**Comment by zoffixznet** [comment]
:+1: 

> This will make some people mad seeing a lot of deprecated call.

Well, 'tis better to do it before Christmas.

File: lib/DBDish/SQLite/Native.pm6 [hunk 1]
```diff
@@ -92,7 +101,7 @@ sub sqlite3_libversion_number() returns int32 is native(LIB) is export { ... };
 sub sqlite3_bind_blob(OpaquePointer $stmt, int32, OpaquePointer, int32, OpaquePointer) returns int32 is native(LIB) is export { ... };
 sub sqlite3_bind_double(OpaquePointer $stmt, int32, num64) returns int32 is native(LIB) is export { ... };
 sub sqlite3_bind_int64(OpaquePointer $stmt, int32, int64) returns int32 is native(LIB) is export { ... };
-sub sqlite3_bind_null(OpaquePointer $stmt, int32) returns Int is native(LIB) is export { ... };
+sub sqlite3_bind_null(OpaquePointer $stmt, int32) returns int32 is native(LIB) is export { ... };
```
Comment (abraxxa) [inline lib/DBDish/SQLite/Native.pm6]: That should be a separate fix/commit!

**Comment by moritz** [comment]
:-1:

We can have a modern interface without breaking backwards compatibility. As long as we don't have MAJOR pain with supporting the old interface, we should continue to do so.

**Comment by ab5tract** [comment]
It seems to me that what we really need is a completely new module and
leave DBIish to be what it is.

Why you would also make this change and then still use smashcase for method
names is also beyond me..

File: lib/DBDish/SQLite/Native.pm6 (hunk 1 above)
Comment (Skarsnik) [inline lib/DBDish/SQLite/Native.pm6]: I just committed this change directly.

**Comment by Skarsnik** [comment]
Notice that I did not change/remove anything that was already here. I just added the DEPRECATED trait on it. Everything will still working as usual
What do you call smashcase? And it's just name, it can be changed.

And the fetch* is a big mess. Some are defined in the role, some backend redefine a part of it for... no reason.

**Comment by ab5tract** [comment]
On Tue, Dec 1, 2015 at 6:52 PM, Skarsnik <notifications@github.com> wrote:

> Notice that I did not change/remove anything that was already here. I just
> added the DEPRECATED trait on it. Everything will still working as usual
> What do you call smashcase? And it's just name, it can be changed.
>
'allrows' is an example of smashcase. You are right that it is just a name,
I didn't mean to be too harsh.


> And the fetch* is a big mess. Some are defined in the role, some backend
> redefine a part of it for... no reason.
>

Which is another good argument to start a new db library from scratch, imo.
I don't mean this as a knock on DBIish, only that I think Perl 6 in 2015
might deserve something that does not already have backwards compatibility
concerns.

**Comment by moritz** [comment]
@Skarsnik there's no use deprecating stuff unless you want to remove it eventually. When you do, you're breaking backwards compatability. (And throwing deprecations warning on code that didn't so before is also breaking backwards compatibility, though in a much milder form).

If the *implementation* of the fetch* methods is messy, we can think about cleaning them up. Provide a bare minimum that a driver must implement, and implement the rest of them in terms of that minimal subset. That can be done in a common row.

If nobody else does, I'll look into it when I get around to it.

**Comment by Skarsnik** [comment]
I don't get what you call backwards compatiblity? For me it's clear that nothing is really fixed with perl6, especially the ecosystem that has lot of 5+ years old stuff never updated. I think with "perl 6 1.0' not being here, it's still the right moment to throw away old thing.

**Comment by moritz** [comment]
A few weeks before the stable release is actually a bad time to throw out old APIs, because the ecosystem typically doesn't catch up in such a short time span.

I still haven't heard a good reason for throwing out the old APIs, so I remain convinced we shouldn't.

**Comment by Skarsnik** [comment]
Ok, maybe I can do another commit without the deprecated trait, but I am not sure how to do that with git. But with completing the doc like saying fetch* are more perl5 DBI and are not encouraged to be used?

I still think it's an interesting discution

**Comment by moritz** [comment]
I'm fine with docs that encourage the new API and discourage the old one.

As for the git part, please take a look at http://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits (or I can help you on IRC, or #git on freenode can).

**Comment by Skarsnik** [comment]
Hm, I need someone with a working mysql server to test the mysql part. It does not want to work on my devel environnement.

**Comment by zoffixznet** [comment]
> Hm, I need someone with a working mysql server to test the mysql part. It does not want to work on my devel environnement.

On my box, the tests freeze after showing this error:

```bash
zoffix@:/tmp/tmp.qPVungYDi0$ git clone https://github.com/Skarsnik/DBIish.git .
Cloning into '.'...
remote: Counting objects: 1481, done.
remote: Compressing objects: 100% (61/61), done.
remote: Total 1481 (delta 25), reused 0 (delta 0), pack-reused 1420
Receiving objects: 100% (1481/1481), 286.06 KiB, done.
Resolving deltas: 100% (760/760), done.
zoffix@:/tmp/tmp.qPVungYDi0$ prove -e "perl6 -Ilib" -vlr t
t/01-connectconfig-pg.t ... 
ok 1 - False PGDATABASE dies ()
# Using PGDATABASE: (none)
ok 2 - database matches PGDATABASE
ok 3 - True PGDATABASE lives ()
ok 4 - database matches PGDATABASE
ok 5 - user matches PGUSER
ok 6 - password matches PGPASSWORD
ok 7 - host is localhost
ok 8 - port is 5432
ok 9 - True PGDATABASE lives ()
ok 10 - database matches PGDATABASE
ok 11 - user matches PGUSER
ok 12 - password matches PGPASSWORD
ok 13 - host is localhost
ok 14 - port is 5432
ok 15 - True PGDATABASE lives ()
1..15
ok
t/05-mock.t ............... 
1..9
===SORRY!=== Error while compiling /tmp/tmp.qPVungYDi0/lib/DBDish/TestMock/Connection.pm6
Invalid typename 'DBDish::Role::Connection'
at /tmp/tmp.qPVungYDi0/lib/DBDish/TestMock/Connection.pm6:6
------> Connection does DBDish::Role::Connection⏏;


^C
zoffix@:/tmp/tmp.qPVungYDi0$ perl6 -v
This is rakudo version 2015.11-311-gf94c31e built on MoarVM version 2015.11-21-g469ba64 implementing Perl v6.b.
zoffix@:/tmp/tmp.qPVungYDi0$ uname -a
Linux  3.2.0-4-686-pae #1 SMP Debian 3.2.51-1 i686 GNU/Linux
zoffix@:/tmp/tmp.qPVungYDi0$
```

**Comment by Skarsnik** [comment]
Alright, finally it work.

I have still a question (I will fill an issue for a it). I try to have row/allrow typed value.
It mean etheir: Int, Str, Num (and Bool for Pg). The Num will probably be rounded in case of too much decimal. DB NUMERIC are mapped to Num but someone pointed out they can be huge. I map then to Real in Pg, but I have no idea what it's really is.

It's probably not a big deal in most of the case so I want to say to still keep typed value. But I can probably add a `:raw `adverb that tell DBIish to ignore field type and add a colomn_rawtype (the native type value returned by the driver) so a highter level API can do the proper thing.

**Comment by abraxxa** [comment]
My work on DBDish::Oracle progressed a lot today and I return VARCHAR2 as Str, NUMBER depending on decimal places as Real or Int and INTEGER as Int.

**Comment by Skarsnik** [comment]
hm, all the fetch* returns everything as Str in the three current drivers. You should probably keep it like that. But keep the typed part for this code.

**Comment by abraxxa** [comment]
OCI only returns typed values. I would have to stringify them which doesn't make sense as a user can do that if (s)he wants to.

File: README.pod [hunk 2]
```diff
@@ -63,9 +63,27 @@ a subset of the functionality).
...
-methods on DBD::StatementHandle that returns a hash with Perl6 typed value by oposition
-to everything in Str (implemented only in Pg for now).
+=head2 Fetching data
+
+It provide nearly all the perl5 DBI fetch* method to fetch values.
+However it's recommanded to use the C<row> and C<allrows> method
+
+=head3 row
+
+C<row> take the <hash> adverbe if you to have the values on a hash form instead of a plain array
```
Comment (vendethiel) [inline README.pod]: Typo "adverbe"

**Pull request merged**